### PR TITLE
TVIST1-545: Document changes

### PR DESCRIPTION
**Pull request by jekuaitk**
* Forces same directory for all documents
* Changes all downloading of documents to view
* Adds `originalFilename` and `path` property to `Document` entity

https://jira.itkdev.dk/browse/TVIST1-545

### CODE REVIEW

File: src/Controller/DecisionController.php [hunk 1]
```diff
@@ -63,7 +63,7 @@ public function create(CaseEntity $case, DigitalPostHelper $digitalPostHelper, D
 
         $caseDocuments = $documentRepository->getAvailableDocumentsForCase($case);
 
-        $documentUploader->specifyDirectory('/case_documents/');
+//        $documentUploader->specifyDirectory('/case_documents/');
```
Comment (rimi-itk) [inline src/Controller/DecisionController.php:66]: Remove this line.

File: src/Controller/DocumentController.php [hunk 2]
```diff
@@ -93,6 +91,7 @@ public function create(CaseEntity $case, Request $request): Response
 
                 // Set filename, document name, creator and case
                 $document->setFilename($newFilename);
+                $document->setPath($this->documentUploader->getUploadDocumentDirectory().'/'.$newFilename);
```
Comment (rimi-itk) [inline src/Controller/DocumentController.php:94]: I think the DocumentUploader should handle this. Where is actually stores files should be kept inside the uploader.

File: src/Entity/Document.php [hunk 3]
```diff
@@ -63,6 +63,11 @@ class Document implements LoggableEntityInterface
      */
     private $hearingPost;
 
+    /**
+     * @ORM\Column(type="string", length=255)
+     */
+    private $path;
```
Comment (rimi-itk) [inline src/Entity/Document.php:69]: A comment on what `path` actually contains could come in handy at some point.

**Review by rimi-itk — CHANGES_REQUESTED** [review]

File: src/Controller/DecisionController.php (hunk 1 above)
Comment (jekuaitk) [reply src/Controller/DecisionController.php:66]: Removed.

File: src/Controller/DocumentController.php (hunk 2 above)
Comment (jekuaitk) [reply src/Controller/DocumentController.php]: Added `getFilepathFromProjectDirectory($filename)` method to `DocumentUploader`.

File: src/Entity/Document.php (hunk 3 above)
Comment (jekuaitk) [reply src/Entity/Document.php]: Added comment.

File: src/Controller/AgendaBroadcastController.php [hunk 4]
```diff
@@ -94,13 +94,13 @@ public function broadcastAgenda(Agenda $agenda, CprHelper $cprHelper, DigitalPos
...
 
             // Move file
-            $documentUploader->specifyDirectory('/case_documents/');
             $updatedFileName = $documentUploader->uploadFile($fileName);
 
             // Create document
             $document = new Document();
             $document->setFilename($updatedFileName);
             $document->setDocumentName($agendaBroadcast->getTitle());
+            $document->setPath($documentUploader->getFilepathFromProjectDirectory($updatedFileName));
```
[the remark this answers is not in the document]
Comment (rimi-itk) [reply src/Controller/AgendaBroadcastController.php:103]: Additional question: Is the `TranslatableMessage` set as the document type automagically converted to a string?

**Review by rimi-itk — CHANGES_REQUESTED** [review]
Looks good. A suggestion for a refactoring added.

File: src/Controller/AgendaBroadcastController.php (hunk 4 above)
[the remark this answers is not in the document]
Comment (jekuaitk) [reply src/Controller/AgendaBroadcastController.php]: Refactored in 1770e2f479929c35ba3f76879cb0030d7bdca835!

File: src/Controller/AgendaBroadcastController.php [hunk 5]
```diff
@@ -96,26 +95,20 @@ public function broadcastAgenda(Agenda $agenda, CprHelper $cprHelper, DigitalPos
...
-            $document->setPath($documentUploader->getFilepathFromProjectDirectory($updatedFileName));
-
             /** @var User $user */
             $user = $this->getUser();
-            $document->setUploadedBy($user);
-            $document->setType(new TranslatableMessage('Agenda broadcast', [], 'agenda'));
 
-            $entityManager->persist($document);
+            // Create document
+            $newDocument = $documentUploader->createDocumentFromPath($updatedFileName, $user, $agendaBroadcast->getTitle(), 'Agenda broadcast');
```
Comment (rimi-itk) [inline src/Controller/AgendaBroadcastController.php:102]: Why is it called “newDocument”? Previously, we didn't use `$newDocument = new Document()` even though that document was also new.

Comment (rimi-itk) [inline src/Controller/AgendaBroadcastController.php:102]: Will it make sense to let the document uploader set the user on the document?

File: src/Service/DocumentUploader.php [hunk 6]
```diff
@@ -38,6 +39,41 @@ public function __construct(SluggerInterface $slugger, string $uploadDocumentDir
...
+        $document->setUploadedBy($user);
+        $document->setType($documentType);
+
+        return $document;
+    }
+
+    /**
+     * Creates, uploads and returns new document from a file.
+     */
+    public function createDocumentFromFile(UploadedFile $file, User $user, string $documentName, string $documentType): Document
```
Comment (rimi-itk) [inline src/Service/DocumentUploader.php:61]: I think this function should be called `createDocumentFromUploadedFile` to make it clear what is does (I was confused about the difference between `createDocumentFromPath` and `createDocumentFromFile` before I came here).

File: src/Service/DocumentUploader.php [hunk 7]
```diff
@@ -38,6 +39,41 @@ public function __construct(SluggerInterface $slugger, string $uploadDocumentDir
...
+    {
+        $document = new Document();
+
+        $document->setDocumentName($documentName);
+        $document->setType($documentType);
+        $document->setOriginalFileName($file->getClientOriginalName());
+        $document->setUploadedBy($user);
+
+        $newFileName = $this->upload($file);
```
Comment (rimi-itk) [inline src/Service/DocumentUploader.php:70]: You could set this at the top of the function body and then return the result of calling `createDocumentFromPath`:

```php
$path = $this->upload($file);
$document = $this->createDocumentFromPath($path, …);
$document->setOriginalFileName($file->getClientOriginalName());

return $document;
```

File: src/Service/DocumentUploader.php [hunk 8]
```diff
@@ -38,6 +39,41 @@ public function __construct(SluggerInterface $slugger, string $uploadDocumentDir
...
+     */
+    public function createDocumentFromPath(string $fileName, User $user, string $documentName, string $documentType): Document
+    {
+        $document = new Document();
+
+        $document->setFilename($fileName);
+        $document->setDocumentName($documentName);
+        $document->setPath($this->getFilepathFromProjectDirectory($fileName));
+        $document->setUploadedBy($user);
+        $document->setType($documentType);
```
Comment (rimi-itk) [inline src/Service/DocumentUploader.php:53]: Should original file name be set here?

**Review by rimi-itk — CHANGES_REQUESTED** [review]

File: src/Service/DocumentUploader.php (hunk 6 above)
Comment (jekuaitk) [reply src/Service/DocumentUploader.php:61]: Changed name as proposed, this does indeed make it much clearer.

File: src/Service/DocumentUploader.php (hunk 7 above)
Comment (jekuaitk) [reply src/Service/DocumentUploader.php:70]: 👍

File: src/Service/DocumentUploader.php (hunk 8 above)
Comment (jekuaitk) [reply src/Service/DocumentUploader.php:53]: Documents created directly via `createDocumentFromPath` all come from a template and therefore don't have an original file name. I will try to make this clearer.

File: src/Controller/AgendaBroadcastController.php (hunk 5 above)
Comment (jekuaitk) [reply src/Controller/AgendaBroadcastController.php:102]: It was initially done in the `DocumentController` as this already had a `$document` variable. 

I will change it back in controllers where this is not the case.

**Review by rimi-itk — APPROVED** [review]
🎉

**Pull request merged**